src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -738,12 +895,164 @@ export class SavedObjectsRepository {
+   * saved object is shared to.
+   */
+  async addToNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsAddToNamespacesOptions = {}
+  ): Promise<{}> {
+    if (!this._allowedTypes.includes(type)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    if (!this._registry.isMultiNamespace(type)) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(`${type} doesn't support namespaces`);
+    }
+
+    if (!Array.isArray(namespaces) || !namespaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'namespaces must be a non-empty array of strings'
+      );
+    }
+
+    const { version, namespace, refresh = DEFAULT_REFRESH_SETTING } = options;
+
+    const rawId = this._serializer.generateRawId(undefined, type, id);
+    const preflightResult = await this.preflightCheckIncludesNamespace(type, id, namespace);
+    const existingNamespaces = getSavedObjectNamespaces(undefined, preflightResult);
+    // there should never be a case where a multi-namespace object does not have any existing namespaces
+    // however, it is a possibility if someone manually modifies the document in Elasticsearch
+    const time = this._getCurrentTime();
+
+    const doc = {
+      updated_at: time,
+      namespaces: existingNamespaces ? unique(existingNamespaces.concat(namespaces)) : namespaces,
+    };
+
+    const updateResponse = await this._writeToCluster('update', {
+      id: rawId,
+      index: this.getIndexForType(type),
+      ...getExpectedVersionProperties(version, preflightResult),
+      refresh,
+      ignore: [404],
+      body: {
+        doc,
+      },
+    });
+
+    if (updateResponse.status === 404) {
+      // see "404s from missing index" above
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    return {};
+  }
+
+  /**
+   * Removes one or more namespaces from a given multi-namespace saved object. If no namespaces remain, the saved object is deleted
+   * entirely. This method and [`addToNamespaces`]{@link SavedObjectsRepository.addToNamespaces} are the only ways to change which Spaces a
+   * multi-namespace saved object is shared to.
+   */
+  async deleteFromNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsDeleteFromNamespacesOptions = {}
+  ): Promise<{}> {
+    if (!this._allowedTypes.includes(type)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    if (!this._registry.isMultiNamespace(type)) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(`${type} doesn't support namespaces`);
+    }
+
+    if (!namespaces || !Array.isArray(namespaces) || !namespaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'namespaces must be a non-empty array of strings'
+      );
+    }
```

Comment:
Apart from `!namespaces.length`, we are performing checks already 'protected' by TS (`namespaces: string[]`). I don't mind that, but in that case shouldn't we also check the type of each element to ensure they are strings? (or even use a `schema.arrayOf(schema.string())` for that?

Reply:
I originally added `!namespaces || !Array.isArray(namespaces)` because we have a similar check for the `references` attribute in the `create`, `update`, and `bulkUpdate` methods. I didn't think too hard about it tbh, just added it here as a basic sanity check. Perhaps that was only in place for the other methods because of legacy (non-TS) consumers?

I am hesitant add extra validation here, that strikes me as unnecessary / a separate concern of route handlers. There is already appropriate string validation in the publicly exposed APIs (`/api/spaces/_share_saved_object_add` and `/api/spaces/_share_saved_object_remove`). 

Perhaps instead, we should remove the `!namespaces || !Array.isArray(namespaces)` checks here since they aren't really necessary?

Reply:
I think these checks were here from the JS time and did not get removed when the file was migrated to TS. I'm fine removing them.

Reply:
Great, updated in 223ecef.